<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Nikhil Vytla">

<title>[RNA Velocity] Trajectory inference and analysis of scRNA-seq data</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="rna_velocity_files/libs/clipboard/clipboard.min.js"></script>
<script src="rna_velocity_files/libs/quarto-html/quarto.js"></script>
<script src="rna_velocity_files/libs/quarto-html/popper.min.js"></script>
<script src="rna_velocity_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="rna_velocity_files/libs/quarto-html/anchor.min.js"></script>
<link href="rna_velocity_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="rna_velocity_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="rna_velocity_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="rna_velocity_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="rna_velocity_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">

  <script>window.backupDefine = window.define; window.define = undefined;</script><script src="https://cdn.jsdelivr.net/npm/katex@0.15.1/dist/katex.min.js"></script>
  <script>document.addEventListener("DOMContentLoaded", function () {
 var mathElements = document.getElementsByClassName("math");
 var macros = [];
 for (var i = 0; i < mathElements.length; i++) {
  var texText = mathElements[i].firstChild;
  if (mathElements[i].tagName == "SPAN") {
   katex.render(texText.data, mathElements[i], {
    displayMode: mathElements[i].classList.contains('display'),
    throwOnError: false,
    macros: macros,
    fleqn: false
   });
}}});
  </script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.15.1/dist/katex.min.css">

</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#setup" id="toc-setup" class="nav-link" data-scroll-target="#setup">Setup</a>
  <ul class="collapse">
  <li><a href="#r" id="toc-r" class="nav-link" data-scroll-target="#r">R</a></li>
  <li><a href="#python" id="toc-python" class="nav-link" data-scroll-target="#python">Python</a></li>
  </ul></li>
  <li><a href="#exploratory-data-analysis" id="toc-exploratory-data-analysis" class="nav-link" data-scroll-target="#exploratory-data-analysis">Exploratory Data Analysis</a>
  <ul class="collapse">
  <li><a href="#load-data" id="toc-load-data" class="nav-link" data-scroll-target="#load-data">Load Data</a></li>
  <li><a href="#downsampling-for-demonstration" id="toc-downsampling-for-demonstration" class="nav-link" data-scroll-target="#downsampling-for-demonstration"><strong>Downsampling for demonstration</strong></a></li>
  </ul></li>
  <li><a href="#basic-rna-velocity-workflow" id="toc-basic-rna-velocity-workflow" class="nav-link" data-scroll-target="#basic-rna-velocity-workflow">Basic RNA Velocity Workflow</a>
  <ul class="collapse">
  <li><a href="#advanced-options" id="toc-advanced-options" class="nav-link" data-scroll-target="#advanced-options"><strong>Advanced options</strong></a></li>
  </ul></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a>
  <ul class="collapse">
  <li><a href="#prior-workshops-websites" id="toc-prior-workshops-websites" class="nav-link" data-scroll-target="#prior-workshops-websites">Prior Workshops / Websites</a></li>
  <li><a href="#papers-articles" id="toc-papers-articles" class="nav-link" data-scroll-target="#papers-articles">Papers / Articles</a></li>
  </ul></li>
  <li><a href="#code-internals" id="toc-code-internals" class="nav-link" data-scroll-target="#code-internals">Code Internals</a>
  <ul class="collapse">
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info">Session Info</a></li>
  </ul></li>
  </ul>
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="rna_velocity.pdf"><i class="bi bi-file-pdf"></i>PDF</a></li></ul></div></nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">[RNA Velocity] Trajectory inference and analysis of scRNA-seq data</h1>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Author</div>
  <div class="quarto-title-meta-heading">Affiliation</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author">Nikhil Vytla </p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            DFCI CCG
          </p>
      </div>
    </div>

<div class="quarto-title-meta">

      
  
    
  </div>
  

</header>

<section id="introduction" class="level1">
<h1>Introduction</h1>
<p>This notebook reviews the use of <strong>RNA velocity</strong> for lineage tracing with a few modern Python/R libraries, including <code>scVelo</code> and <code>velociraptor</code>, which build on original work from <code>velocyto</code> (last maintained in 2017).</p>
<p>What is <strong>RNA velocity</strong>? To understand RNA velocity, we must first recognize two ideas:</p>
<ol type="1">
<li>The RNA metabolic process and its intrinsic dynamics:
<ol type="1">
<li>RNA is synthesized at a rate of <span class="math inline">\alpha</span></li>
<li>RNA is spliced to remove introns and form mature mRNA at rate of <span class="math inline">\beta</span></li>
<li>Mature mRNA is degraded after functioning at a rate of <span class="math inline">\gamma</span></li>
<li>In steady state, the unspliced and spliced RNA are at an equilibrium (newly synthesized RNAs are exactly equal to the spliced RNAs, and similarly also to the degradated RNAs).</li>
<li>Therefore, the break of the equilibrium state between unspliced and spliced RNAs is <em>highly informative</em> and can indicate whether a gene is in the induction or suppression state.</li>
</ol></li>
<li>Unspliced RNA indicates transcriptional speed
<ol type="1">
<li>Usually, the proportion of unspliced RNAs is very low in RNA-seq, especially for protocols with Poly-A enrichment, which is very common.</li>
<li>Namely, in principle, only RNAs reaching 3' of the gene body can be captured.</li>
<li>However, the unspliced RNAs still can be observed at a substantial proportion, usually covering 15-25%. The reason is still highly mysterious, partly biological for co-transcriptional splicing and partly technical for low efficiency on poly-A capturing.</li>
</ol></li>
</ol>
<p>While RNA velocity is still impacted by high signal-to-noise ratio, the ability to automatically detect the trajectory direction continues to draw research efforts to develop more accurate and robust methods.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/rna_velo_process.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">RNA velocity process, adapted from <a href="https://www.nature.com/articles/s41586-018-0414-6">La Manno et al.&nbsp;2018</a> (this paper introduced <code>velocyto</code>).</figcaption>
</figure>
</div>
</section>
<section id="setup" class="level1">
<h1>Setup</h1>
<section id="r" class="level2">
<h2 class="anchored" data-anchor-id="r">R</h2>
<p>This notebook will utilize the R package <code>velociraptor</code>. <code>velociraptor</code> provides a lightweight interface between the Bioconductor <code>SingleCellExperiment</code> data structure and the <code>scVelo</code> Python package for RNA velocity calculations.</p>
<p>The interface is comparable to that of many other <code>SingleCellExperiment</code>-compatible functions, allowing users to plug in RNA velocity calculations into the existing Bioconductor analysis framework.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">"BiocManager"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="fu">c</span>(<span class="st">"scRNAseq"</span>, <span class="st">"velociraptor"</span>, <span class="st">"scuttle"</span>, <span class="st">"scran"</span>), <span class="at">dependencies =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load libraries</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scRNAseq) <span class="co"># data loading</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scuttle)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scran)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(velociraptor)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scater)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="python" class="level2">
<h2 class="anchored" data-anchor-id="python">Python</h2>
<p><code>scVelo</code> is originally built in Python, and by default uses the <code>AnnData</code> format. Please check out the following official Jupyter Notebook tutorials for guidance:</p>
<ul>
<li><p><a href="https://scvelo.readthedocs.io/en/stable/VelocityBasics.html">RNA Velocity Basics</a></p></li>
<li><p><a href="https://scvelo.readthedocs.io/en/stable/DynamicalModeling.html">Dynamical Modeling</a></p></li>
<li><p><a href="https://scvelo.readthedocs.io/en/stable/DifferentialKinetics.html">Differential Kinetics</a></p></li>
</ul>
</section>
</section>
<section id="exploratory-data-analysis" class="level1">
<h1>Exploratory Data Analysis</h1>
<p>To demonstrate, we will use a data set from <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6384825/">Hermann et al.&nbsp;(2018)</a>, provided via the <code>scRNAseq</code> package. This data set contains gene-wise estimates of spliced and unspliced UMI counts for 2,325 mouse spermatogenic cells.</p>
<section id="load-data" class="level2">
<h2 class="anchored" data-anchor-id="load-data">Load Data</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">HermannSpermatogenesisData</span>()</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>sce</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SingleCellExperiment 
dim: 54448 2325 
metadata(0):
assays(2): spliced unspliced
rownames(54448): ENSMUSG00000102693.1 ENSMUSG00000064842.1 ...
  ENSMUSG00000064369.1 ENSMUSG00000064372.1
rowData names(0):
colnames(2325): CCCATACTCCGAAGAG AATCCAGTCATCTGCC ... ATCCACCCACCACCAG
  ATTGGTGGTTACCGAT
colData names(1): celltype
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
</div>
</section>
<section id="downsampling-for-demonstration" class="level2">
<h2 class="anchored" data-anchor-id="downsampling-for-demonstration"><strong>Downsampling for demonstration</strong></h2>
<p>The full data set requires up to 12 GB of memory for the example usage presented in this vignette. For demonstration purposes, we downsample the data set to the first 500 cells. Feel free to skip this downsampling step if you have access to sufficient memory.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> sce[, <span class="dv">1</span><span class="sc">:</span><span class="dv">500</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="basic-rna-velocity-workflow" class="level1">
<h1>Basic RNA Velocity Workflow</h1>
<p>We assume that feature selection has already been performed by the user using any method (see <a href="https://osca.bioconductor.org/feature-selection.html">here</a> for some suggestions). In this case, we will use the variance of log-expressions from <a href="https://bioconductor.org/packages/3.18/scran"><em>scran</em></a> to select the top 2000 genes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">logNormCounts</span>(sce, <span class="at">assay.type=</span><span class="dv">1</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>dec <span class="ot">&lt;-</span> <span class="fu">modelGeneVar</span>(sce)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in regularize.values(x, y, ties, missing(ties), na.rm = na.rm):
collapsing to unique 'x' values</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>top.hvgs <span class="ot">&lt;-</span> <span class="fu">getTopHVGs</span>(dec, <span class="at">n=</span><span class="dv">2000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can plug these choices into the <a href="https://kevinrue.github.io/velociraptor/reference/scvelo.html"><code>scvelo()</code></a> function with our <code>SingleCellExperiment</code> object. By default, <a href="https://kevinrue.github.io/velociraptor/reference/scvelo.html"><code>scvelo()</code></a> uses the steady-state approach to estimate velocities, though the stochastic and dynamical models implemented in <a href="https://pypi.org/project/scvelo">scvelo</a> can also be used by modifying the <code>mode</code> argument.</p>
<p>Note that automatic neighbor calculation is deprecated since scvelo==0.4.0 and will be removed in a future version. Instead, <a href="https://bioconductor.org/packages/3.18/velociraptor"><em>velociraptor</em></a> computes neighbors with Scanpy (as per <a href="https://github.com/theislab/scvelo/blob/f21651c3b122860d8ae6b5a5173f242ba91c8761/scvelo/preprocessing/moments.py#L66">scVelo recommendations</a>), and the number of neighbors should be supplied to <a href="https://scanpy.readthedocs.io/en/stable/api/generated/scanpy.pp.neighbors.html">scanpy.pp.neighbors</a> as demonstrated below.</p>
<p>In particular, the default number of neighbors was 30 for <a href="https://scvelo.readthedocs.io/en/stable/scvelo.pp.moments.html">scvelo.pp.moments</a> while it is 15 for <a href="https://scanpy.readthedocs.io/en/stable/api/generated/scanpy.pp.neighbors.html">scanpy.pp.neighbors</a>. Users should use <code>scvelo.params=list(neighbors=list(n_neighbors=30L)</code> to reproduce earlier results.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>velo.out <span class="ot">&lt;-</span> <span class="fu">scvelo</span>(</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  sce, <span class="at">subset.row=</span>top.hvgs, <span class="at">assay.X=</span><span class="st">"spliced"</span>,</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">scvelo.params=</span><span class="fu">list</span>(<span class="at">neighbors=</span><span class="fu">list</span>(<span class="at">n_neighbors=</span>30L))</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>computing moments based on connectivities
    finished (0:00:00) --&gt; added 
    'Ms' and 'Mu', moments of un/spliced abundances (adata.layers)
computing velocities
    finished (0:00:00) --&gt; added 
    'velocity', velocity vectors for each individual cell (adata.layers)
computing velocity graph (using 1/10 cores)
WARNING: Unable to create progress bar. Consider installing `tqdm` as `pip install tqdm` and `ipywidgets` as `pip install ipywidgets`,
or disable the progress bar using `show_progress_bar=False`.
    finished (0:00:00) --&gt; added 
    'velocity_graph', sparse matrix with cosine correlations (adata.uns)
computing terminal states
    identified 1 region of root cells and 1 region of end points .
    finished (0:00:00) --&gt; added
    'root_cells', root cells of Markov diffusion process (adata.obs)
    'end_points', end points of Markov diffusion process (adata.obs)
--&gt; added 'velocity_length' (adata.obs)
--&gt; added 'velocity_confidence' (adata.obs)
--&gt; added 'velocity_confidence_transition' (adata.obs)</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>velo.out</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SingleCellExperiment 
dim: 2000 500 
metadata(4): neighbors velocity_params velocity_graph
  velocity_graph_neg
assays(6): X spliced ... Mu velocity
rownames(2000): ENSMUSG00000117819.1 ENSMUSG00000081984.3 ...
  ENSMUSG00000022965.8 ENSMUSG00000094660.2
rowData names(4): velocity_gamma velocity_qreg_ratio velocity_r2
  velocity_genes
colnames(500): CCCATACTCCGAAGAG AATCCAGTCATCTGCC ... CACCTTGTCGTAGGAG
  TTCCCAGAGACTAAGT
colData names(7): velocity_self_transition root_cells ...
  velocity_confidence velocity_confidence_transition
reducedDimNames(1): X_pca
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
</div>
<p>In the above call, we use the <code>"spliced"</code> count matrix as a proxy for the typical exonic count matrix. Technically, the latter is not required for the velocity estimation, but <a href="https://pypi.org/project/scvelo">scvelo</a> needs to perform a PCA and nearest neighbors search, and we want to ensure that the neighbors detected inside the function are consistent with the rest of the analysis workflow (performed on the exonic counts).</p>
<p>There are some subtle differences between the spliced count matrix and the typical exonic count matrix - see <a href="https://kevinrue.github.io/velociraptor/reference/scvelo.html"><code>?scvelo</code></a> for some commentary about this - but the spliced counts are generally a satisfactory replacement if the latter is not available.</p>
<p>The <a href="https://kevinrue.github.io/velociraptor/reference/scvelo.html"><code>scvelo()</code></a> function produces a <code>SingleCellExperiment</code> containing all of the outputs of the calculation in Python. Of particular interest is the <code>velocity_pseudotime</code> vector that captures the relative progression of each cell along the biological process driving the velocity vectors.</p>
<p>We can visualize this effect below in a <span class="math inline">t</span>-SNE plot generated by <a href="https://bioconductor.org/packages/3.18/scater"><em>scater</em></a> on the top HVGs.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">100</span>) <span class="co"># for reproducibility</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(sce, <span class="at">subset_row=</span>top.hvgs)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">runTSNE</span>(sce, <span class="at">dimred=</span><span class="st">"PCA"</span>, <span class="at">perplexity =</span> <span class="dv">30</span>)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">$</span>velocity_pseudotime <span class="ot">&lt;-</span> velo.out<span class="sc">$</span>velocity_pseudotime</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTSNE</span>(sce, <span class="at">colour_by=</span><span class="st">"velocity_pseudotime"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="rna_velocity_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>It is also straightforward to embed the velocity vectors into our desired low-dimensional space, as shown below for the <span class="math inline">t</span>-SNE coordinates. This uses a grid-based approach to summarize the per-cell vectors into local representatives for effective visualization.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>embedded <span class="ot">&lt;-</span> <span class="fu">embedVelocity</span>(<span class="fu">reducedDim</span>(sce, <span class="st">"TSNE"</span>), velo.out)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>ℹ Using the 'X' assay as the X matrix</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>computing velocity embedding
    finished (0:00:00) --&gt; added
    'velocity_target', embedded velocity vectors (adata.obsm)</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>grid.df <span class="ot">&lt;-</span> <span class="fu">gridVectors</span>(sce, embedded, <span class="at">use.dimred =</span> <span class="st">"TSNE"</span>)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTSNE</span>(sce, <span class="at">colour_by=</span><span class="st">"velocity_pseudotime"</span>) <span class="sc">+</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_segment</span>(<span class="at">data=</span>grid.df, <span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>start<span class="fl">.1</span>, <span class="at">y=</span>start<span class="fl">.2</span>, </span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">xend=</span>end<span class="fl">.1</span>, <span class="at">yend=</span>end<span class="fl">.2</span>, <span class="at">colour=</span><span class="cn">NULL</span>), <span class="at">arrow=</span><span class="fu">arrow</span>(<span class="at">length=</span><span class="fu">unit</span>(<span class="fl">0.05</span>, <span class="st">"inches"</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="rna_velocity_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>And that's it, really.</p>
<section id="advanced-options" class="level2">
<h2 class="anchored" data-anchor-id="advanced-options"><strong>Advanced options</strong></h2>
<p><a href="https://kevinrue.github.io/velociraptor/reference/scvelo.html"><code>scvelo()</code></a> interally performs a PCA step that we can bypass by supplying our own PC coordinates.</p>
<p>It is often the case that we have already performed PCA in the earlier analysis steps, so we can just re-use those results to:</p>
<ol type="1">
<li>save time and</li>
<li>improve consistency with the other steps.</li>
</ol>
<p>Here, we computed the PCA coordinates in <a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html"><code>runPCA()</code></a> above, so let's just recycle that:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Only setting assay.X= for the initial AnnData creation,</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="co"># it is not actually used in any further steps.</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>velo.out2 <span class="ot">&lt;-</span> <span class="fu">scvelo</span>(sce, <span class="at">assay.X=</span><span class="dv">1</span>, <span class="at">subset.row=</span>top.hvgs, <span class="at">use.dimred=</span><span class="st">"PCA"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>computing moments based on connectivities
    finished (0:00:00) --&gt; added 
    'Ms' and 'Mu', moments of un/spliced abundances (adata.layers)
computing velocities
    finished (0:00:00) --&gt; added 
    'velocity', velocity vectors for each individual cell (adata.layers)
computing velocity graph (using 1/10 cores)
    finished (0:00:00) --&gt; added 
    'velocity_graph', sparse matrix with cosine correlations (adata.uns)
computing terminal states
    identified 3 regions of root cells and 1 region of end points .
    finished (0:00:00) --&gt; added
    'root_cells', root cells of Markov diffusion process (adata.obs)
    'end_points', end points of Markov diffusion process (adata.obs)
--&gt; added 'velocity_length' (adata.obs)
--&gt; added 'velocity_confidence' (adata.obs)
--&gt; added 'velocity_confidence_transition' (adata.obs)</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>velo.out2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SingleCellExperiment 
dim: 2000 500 
metadata(4): neighbors velocity_params velocity_graph
  velocity_graph_neg
assays(6): X spliced ... Mu velocity
rownames(2000): ENSMUSG00000117819.1 ENSMUSG00000081984.3 ...
  ENSMUSG00000022965.8 ENSMUSG00000094660.2
rowData names(4): velocity_gamma velocity_qreg_ratio velocity_r2
  velocity_genes
colnames(500): CCCATACTCCGAAGAG AATCCAGTCATCTGCC ... CACCTTGTCGTAGGAG
  TTCCCAGAGACTAAGT
colData names(7): velocity_self_transition root_cells ...
  velocity_confidence velocity_confidence_transition
reducedDimNames(1): X_pca
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
</div>
<p><code>velociraptor</code> also provides an option to use the <a href="https://pypi.org/project/scvelo">scvelo</a> pipeline without modification, i.e., relying on their normalization and feature selection.</p>
<p>This sacrifices consistency with other Bioconductor workflows but enables perfect mimicry of a pure Python-based analysis. In this case, arguments like <code>subset.row=</code> are simply ignored.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>velo.out3 <span class="ot">&lt;-</span> <span class="fu">scvelo</span>(sce, <span class="at">assay.X=</span><span class="dv">1</span>, <span class="at">use.theirs=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>WARNING: Did not normalize X as it looks processed already. To enforce normalization, set `enforce=True`.
WARNING: Did not normalize spliced as it looks processed already. To enforce normalization, set `enforce=True`.
WARNING: Did not normalize unspliced as it looks processed already. To enforce normalization, set `enforce=True`.
Logarithmized X.
computing moments based on connectivities
    finished (0:00:00) --&gt; added 
    'Ms' and 'Mu', moments of un/spliced abundances (adata.layers)
computing velocities
    finished (0:00:01) --&gt; added 
    'velocity', velocity vectors for each individual cell (adata.layers)
computing velocity graph (using 1/10 cores)
    finished (0:00:00) --&gt; added 
    'velocity_graph', sparse matrix with cosine correlations (adata.uns)
computing terminal states
    identified 1 region of root cells and 1 region of end points .
    finished (0:00:00) --&gt; added
    'root_cells', root cells of Markov diffusion process (adata.obs)
    'end_points', end points of Markov diffusion process (adata.obs)
--&gt; added 'velocity_length' (adata.obs)
--&gt; added 'velocity_confidence' (adata.obs)
--&gt; added 'velocity_confidence_transition' (adata.obs)</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>velo.out3</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SingleCellExperiment 
dim: 54448 500 
metadata(5): log1p neighbors velocity_params velocity_graph
  velocity_graph_neg
assays(6): X spliced ... Mu velocity
rownames(54448): ENSMUSG00000102693.1 ENSMUSG00000064842.1 ...
  ENSMUSG00000064369.1 ENSMUSG00000064372.1
rowData names(4): velocity_gamma velocity_qreg_ratio velocity_r2
  velocity_genes
colnames(500): CCCATACTCCGAAGAG AATCCAGTCATCTGCC ... CACCTTGTCGTAGGAG
  TTCCCAGAGACTAAGT
colData names(11): initial_size_unspliced initial_size_spliced ...
  velocity_confidence velocity_confidence_transition
reducedDimNames(1): X_pca
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
</div>
<p>Advanced users can tinker with the settings of individual <a href="https://pypi.org/project/scvelo">scvelo</a> steps by setting named lists of arguments in the <code>scvelo.params</code> argument.</p>
<p>For example, to tinker with the behavior of the <code>recover_dynamics</code> step, we could do:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>velo.out4 <span class="ot">&lt;-</span> <span class="fu">scvelo</span>(sce, <span class="at">assay.X=</span><span class="dv">1</span>, <span class="at">subset.row=</span>top.hvgs,</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">scvelo.params=</span><span class="fu">list</span>(<span class="at">recover_dynamics=</span><span class="fu">list</span>(<span class="at">max_iter=</span><span class="dv">20</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>computing moments based on connectivities
    finished (0:00:00) --&gt; added 
    'Ms' and 'Mu', moments of un/spliced abundances (adata.layers)
computing velocities
    finished (0:00:00) --&gt; added 
    'velocity', velocity vectors for each individual cell (adata.layers)
computing velocity graph (using 1/10 cores)
    finished (0:00:00) --&gt; added 
    'velocity_graph', sparse matrix with cosine correlations (adata.uns)
computing terminal states
    identified 2 regions of root cells and 1 region of end points .
    finished (0:00:00) --&gt; added
    'root_cells', root cells of Markov diffusion process (adata.obs)
    'end_points', end points of Markov diffusion process (adata.obs)
--&gt; added 'velocity_length' (adata.obs)
--&gt; added 'velocity_confidence' (adata.obs)
--&gt; added 'velocity_confidence_transition' (adata.obs)</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>velo.out4</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SingleCellExperiment 
dim: 2000 500 
metadata(4): neighbors velocity_params velocity_graph
  velocity_graph_neg
assays(6): X spliced ... Mu velocity
rownames(2000): ENSMUSG00000117819.1 ENSMUSG00000081984.3 ...
  ENSMUSG00000022965.8 ENSMUSG00000094660.2
rowData names(4): velocity_gamma velocity_qreg_ratio velocity_r2
  velocity_genes
colnames(500): CCCATACTCCGAAGAG AATCCAGTCATCTGCC ... CACCTTGTCGTAGGAG
  TTCCCAGAGACTAAGT
colData names(7): velocity_self_transition root_cells ...
  velocity_confidence velocity_confidence_transition
reducedDimNames(1): X_pca
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
</div>
</section>
</section>
<section id="references" class="level1">
<h1>References</h1>
<section id="prior-workshops-websites" class="level2">
<h2 class="anchored" data-anchor-id="prior-workshops-websites">Prior Workshops / Websites</h2>
<ul>
<li><a href="https://kevinrue.github.io/velociraptor/articles/velociraptor.html">Velociraptor R documentation</a></li>
</ul>
<!-- -->
<ul>
<li><p><a href="https://statbiomed.github.io/SingleCell-Workshop-2021">HKU Single-cell Workshop 2021</a> (Section 3)</p></li>
<li><p><a href="https://singlecellanalysistutorial.readthedocs.io/en/latest/index.html">New Academic Year “Cell Diverse” Single Cell Analysis Technique Workshop (October 2019, Nakoto Lab, UTokyo)</a></p></li>
</ul>
</section>
<section id="papers-articles" class="level2">
<h2 class="anchored" data-anchor-id="papers-articles">Papers / Articles</h2>
<p>Hermann, Brian P, Keren Cheng, Anukriti Singh, Lorena Roa-De La Cruz, Kazadi N Mutoji, I-Chung Chen, Heidi Gildersleeve, et al.&nbsp;2018. "The Mammalian Spermatogenesis Single-Cell Transcriptome, from Spermatogonial Stem Cells to Spermatids." <em>Cell Rep.</em> 25: 1650–1667.e8.</p>
</section>
</section>
<section id="code-internals" class="level1">
<h1>Code Internals</h1>
<section id="session-info" class="level2">
<h2 class="anchored" data-anchor-id="session-info">Session Info</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.4.1 (2024-06-14)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.5

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] scater_1.32.1               ggplot2_3.5.1              
 [3] velociraptor_1.14.3         scran_1.32.0               
 [5] scuttle_1.14.0              scRNAseq_2.18.0            
 [7] SingleCellExperiment_1.26.0 SummarizedExperiment_1.34.0
 [9] Biobase_2.64.0              GenomicRanges_1.56.1       
[11] GenomeInfoDb_1.40.1         IRanges_2.38.1             
[13] S4Vectors_0.42.1            BiocGenerics_0.50.0        
[15] MatrixGenerics_1.16.0       matrixStats_1.3.0          

loaded via a namespace (and not attached):
  [1] rstudioapi_0.16.0         jsonlite_1.8.8           
  [3] magrittr_2.0.3            ggbeeswarm_0.7.2         
  [5] GenomicFeatures_1.56.0    gypsum_1.0.1             
  [7] farver_2.1.2              rmarkdown_2.27           
  [9] BiocIO_1.14.0             zlibbioc_1.50.0          
 [11] vctrs_0.6.5               memoise_2.0.1            
 [13] Rsamtools_2.20.0          DelayedMatrixStats_1.26.0
 [15] RCurl_1.98-1.16           htmltools_0.5.8.1        
 [17] S4Arrays_1.4.1            AnnotationHub_3.12.0     
 [19] curl_5.2.1                BiocNeighbors_1.22.0     
 [21] Rhdf5lib_1.26.0           SparseArray_1.4.8        
 [23] rhdf5_2.48.0              alabaster.base_1.4.2     
 [25] htmlwidgets_1.6.4         basilisk_1.16.0          
 [27] alabaster.sce_1.4.0       httr2_1.0.2              
 [29] cachem_1.1.0              GenomicAlignments_1.40.0 
 [31] igraph_2.0.3              lifecycle_1.0.4          
 [33] pkgconfig_2.0.3           rsvd_1.0.5               
 [35] Matrix_1.7-0              R6_2.5.1                 
 [37] fastmap_1.2.0             GenomeInfoDbData_1.2.12  
 [39] digest_0.6.36             colorspace_2.1-1         
 [41] AnnotationDbi_1.66.0      dqrng_0.4.1              
 [43] irlba_2.3.5.1             ExperimentHub_2.12.0     
 [45] RSQLite_2.3.7             beachmat_2.20.0          
 [47] labeling_0.4.3            filelock_1.0.3           
 [49] fansi_1.0.6               httr_1.4.7               
 [51] abind_1.4-5               compiler_4.4.1           
 [53] withr_3.0.1               bit64_4.0.5              
 [55] BiocParallel_1.38.0       viridis_0.6.5            
 [57] DBI_1.2.3                 HDF5Array_1.32.0         
 [59] alabaster.ranges_1.4.2    alabaster.schemas_1.4.0  
 [61] rappdirs_0.3.3            DelayedArray_0.30.1      
 [63] rjson_0.2.21              bluster_1.14.0           
 [65] tools_4.4.1               vipor_0.4.7              
 [67] beeswarm_0.4.0            glue_1.7.0               
 [69] restfulr_0.0.15           rhdf5filters_1.16.0      
 [71] grid_4.4.1                Rtsne_0.17               
 [73] cluster_2.1.6             generics_0.1.3           
 [75] gtable_0.3.5              ensembldb_2.28.0         
 [77] BiocSingular_1.20.0       ScaledMatrix_1.12.0      
 [79] metapod_1.12.0            utf8_1.2.4               
 [81] XVector_0.44.0            ggrepel_0.9.5            
 [83] BiocVersion_3.19.1        pillar_1.9.0             
 [85] limma_3.60.4              dplyr_1.1.4              
 [87] BiocFileCache_2.12.0      lattice_0.22-6           
 [89] rtracklayer_1.64.0        bit_4.0.5                
 [91] tidyselect_1.2.1          locfit_1.5-9.10          
 [93] Biostrings_2.72.1         knitr_1.48               
 [95] gridExtra_2.3             ProtGenerics_1.36.0      
 [97] edgeR_4.2.1               xfun_0.46                
 [99] statmod_1.5.0             UCSC.utils_1.0.0         
[101] lazyeval_0.2.2            yaml_2.3.10              
[103] evaluate_0.24.0           codetools_0.2-20         
[105] tibble_3.2.1              alabaster.matrix_1.4.2   
[107] BiocManager_1.30.23       cli_3.6.3                
[109] reticulate_1.38.0         munsell_0.5.1            
[111] zellkonverter_1.14.1      Rcpp_1.0.13              
[113] dir.expiry_1.12.0         dbplyr_2.5.0             
[115] png_0.1-8                 XML_3.99-0.17            
[117] parallel_4.4.1            blob_1.2.4               
[119] basilisk.utils_1.16.0     AnnotationFilter_1.28.0  
[121] sparseMatrixStats_1.16.0  bitops_1.0-8             
[123] viridisLite_0.4.2         alabaster.se_1.4.1       
[125] scales_1.3.0              crayon_1.5.3             
[127] rlang_1.1.4               cowplot_1.1.3            
[129] KEGGREST_1.44.1          </code></pre>
</div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>